### PR TITLE
Redesign

### PR DESCRIPTION
**Pull request by vudaltsov**

### CODE REVIEW

File: src/types.php [hunk 1]
```diff
@@ -6,54 +6,54 @@
...
+    case object = 'object';
+    case positiveInt = 'positive-int';
+    case resource = 'resource';
+    case scalar = 'scalar';
+    case string = 'string';
+    case true = 'true';
+    case truthyString = 'truthy-string';
+    case void = 'void';
+
+    public const nonFalsyString = self::truthyString;
```
Comment (klimick) [inline src/types.php]: Maybe just remove it?

File: src/types.php [hunk 2]
```diff
@@ -409,109 +429,60 @@ public static function template(string $name, AtMethod|AtClass|AtFunction $decla
...
-const __callable = new CallableType();
+    public function accept(TypeVisitor $visitor): mixed
+    {
+        return match ($this) {
+            self::array => $visitor->visitArray(self::arrayKey, self::mixed),
+            self::arrayKey => $visitor->visitUnion([self::int, self::string]),
+            self::bool => $visitor->visitBool(),
+            self::callable => $visitor->visitCallable([], self::mixed),
+            self::classString => $visitor->visitClassString(),
+            self::Closure => $visitor->visitNamedObject(\Closure::class, []),
```
Comment (klimick) [inline src/types.php]: `$visitor->visitClosure([], self::mixed)` ?

**Review by klimick — CHANGES_REQUESTED** [review]

File: src/types.php (hunk 1 above)
Comment (vudaltsov) [reply src/types.php]: Sure.

File: src/types.php (hunk 2 above)
Comment (vudaltsov) [reply src/types.php]: Well, this is debatable. Most of the time when Closure is used without params and return type, it is meant as a named object. In native types, for instance. On the other hand it makes sense to choose stricter type whenever possible. I will do `visitClosure`, agree.

**Review by klimick — APPROVED** [review]

**Pull request merged**